server/apiv2/handlers/stores.go

```diff
@@ -0,0 +1,180 @@
+// @Failure 500 {string} string "PD server failed to proceed the request."
+// @Router /stores/{id} [get]
+func GetStoreByID() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		rc := c.MustGet("cluster").(*cluster.RaftCluster)
+		idParam := c.Param("id")
+		id, err := strconv.ParseUint(idParam, 10, 64)
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusBadRequest, errs.ErrStrconvParseUint.Wrap(err).FastGenWithCause().Error())
+			return
+		}
+		store := rc.GetStore(id)
+		if store == nil {
+			c.AbortWithStatus(http.StatusNotFound)
+			return
+		}
+
+		storeInfo := newStoreInfo(rc.GetOpts().GetScheduleConfig(), store)
+		c.IndentedJSON(http.StatusOK, storeInfo)
+	}
+}
+
+// DeleteStoreByID will delete the store according to the given ID.
+func DeleteStoreByID() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		rc := c.MustGet("cluster").(*cluster.RaftCluster)
+		idParam := c.Param("id")
+		id, err := strconv.ParseUint(idParam, 10, 64)
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusBadRequest, errs.ErrStrconvParseUint.Wrap(err).FastGenWithCause().Error())
+			return
+		}
+
+		var force bool
+		forceQuery, exist := c.GetQuery("force")
+		if exist && forceQuery != "" {
+			force, err = strconv.ParseBool(forceQuery)
+			if err != nil {
+				c.AbortWithStatusJSON(http.StatusBadRequest, errs.ErrStrconvParseBool.Wrap(err).FastGenWithCause().Error())
+				return
+			}
+		}
+
+		err = rc.RemoveStore(id, force)
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusBadRequest, err.Error())
+			return
+		}
+
+		c.JSON(http.StatusOK, nil)
+	}
+}
+
+type updateStoresParams struct {
+	NodeState    string  `json:"node_state"`
+	LeaderWeight float64 `json:"leader_weight"`
+	RegionWeight float64 `json:"region_weight"`
+}
+
+// UpdateStoreByID will delete the store according to the given ID.
+func UpdateStoreByID() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		rc := c.MustGet("cluster").(*cluster.RaftCluster)
+		idParam := c.Param("id")
+		id, err := strconv.ParseUint(idParam, 10, 64)
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusBadRequest, errs.ErrStrconvParseUint.Wrap(err).FastGenWithCause().Error())
+			return
+		}
+
+		var p updateStoresParams
+		if err := c.BindJSON(&p); err != nil {
+			c.AbortWithStatusJSON(http.StatusBadRequest, errs.ErrBindJSON.Wrap(err).GenWithStackByCause())
+			return
+		}
+
+		if p.LeaderWeight < 0 || p.RegionWeight < 0 {
+			c.AbortWithStatus(http.StatusBadRequest)
+			return
+		}
+
+		if err := rc.SetStoreWeight(id, p.LeaderWeight, p.RegionWeight); err != nil {
+			if errors.Is(err, errs.ErrStoreNotFound) {
+				c.AbortWithStatus(http.StatusNotFound)
+				return
+			}
+			c.AbortWithStatusJSON(http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		if p.NodeState != "" {
+			switch p.NodeState {
+			case metapb.NodeState_Serving.String():
+				if err = rc.UpStore(id); errors.Is(err, errs.ErrStoreNotFound) {
+					c.AbortWithStatus(http.StatusNotFound)
```

Comment:
Should we consider the situation that  `SetStoreWeight` success while `UpStore` failed, I think this API should be automic.

Reply:
That is indeed a problem, I don't have a better idea to make it atomic since it has multiple actions. Unless we revert the change if one of these steps fails, but it's too complex I think.